### ISSUE TITLE
Where is the positional encoding in the Dilated Transformer?

### ISSUE THREAD

**Issue by YiwenShan** [Wang-XIaoDingdd/LGBPN#7]
class Attention_dilated(nn.Module):
    def __init__(self, dim, num_heads, bias, stride):
        super(Attention_dilated, self).__init__()
        self.num_heads = num_heads
        self.temperature = nn.Parameter(torch.ones(num_heads, 1, 1))

        self.qkv = nn.Conv2d(dim, dim * 3, kernel_size=1, bias=bias) # 1*1 维度升3倍, 均分成q,k,v
        self.qkv_dwconv = nn.Conv2d(dim * 3, dim * 3, kernel_size=3, stride=1, padding=stride, groups=dim * 3, bias=bias, dilation=stride) # [3*dim,1, 3,3](channel-wise)  [B,3dim,H,W]->[B,3dim,H,W]
        self.project_out = nn.Conv2d(dim, dim, kernel_size=1, bias=bias)

    def forward(self, x): # x:[B,D',H,W]
        b, c, h, w = x.shape

        qkv = self.qkv_dwconv(self.qkv(x)) # [B,D',H,W] --qkv(1*1卷积)--> [B,3D',H,W] --dwconv(k=3,s=1,p=3,dil=3,groups=3D')--> [B,3D', H,W]
        q, k, v = qkv.chunk(3, dim=1) # q,k,v:[B,D',H,W]  chunk(3,dim=1)沿第1维将[B,3D',H,W]等分成3份

        q = rearrange(q, 'b (head c) h w -> b head c (h w)', head=self.num_heads) # [B,D',H,W] ->[B, hd, D'/hd, H*W]
        k = rearrange(k, 'b (head c) h w -> b head c (h w)', head=self.num_heads) # [B,D',H,W] ->[B, hd, D'/hd, H*W]
        v = rearrange(v, 'b (head c) h w -> b head c (h w)', head=self.num_heads) # [B,D',H,W] ->[B, hd, D'/hd, H*W]

        q = torch.nn.functional.normalize(q, dim=-1) # [B, hd, D'/hd, H*W]
        k = torch.nn.functional.normalize(k, dim=-1) # [B, hd, D'/hd, H*W]

        attn = (q @ k.transpose(-2, -1)) * self.temperature # [B, hd, D'/hd, H*W] * [B, hd, H*W, D'/hd] = [B,hd, D'/hd, D'/hd] * [hd,1,1] = [B,hd, D'/hd, D'/hd]
        attn = attn.softmax(dim=-1) # [B,hd, D'/hd, D'/hd]

        out = (attn @ v) # [B,hd, D'/hd, D'/hd] * [B,hd, D'/hd, H*W] = [B,hd, D'/hd, H*W]

        out = rearrange(out, 'b head c (h w) -> b (head c) h w', head=self.num_heads, h=h, w=w) # [B,hd, D'/hd, H*W] ->[B, D', H,W]

        out = self.project_out(out) # [B,D',H,W] ->还是[B,D',H,W]
        return out

Where is the positional encoding?
